engine/src/engine.rs

```diff
@@ -431,13 +431,12 @@ impl<'env, I: IO + Copy, E: Env> Engine<'env, I, E> {
             .checked_mul(effective_gas_price)
```

Comment:
if `effective_gas_price` is zero, we can do a bit optimize it:
- `prepaid_amount` - in that case always = 0, we don't need `mul` it
- `new_balance` is the same as `balance`, we don't need `sub` it
- we don't need cache it, as the balance was not changed

Reply:
I don't think that checks are much much cheaper than `mul` and `sub`. IMO, it's not worse it.

Reply:
And the goal of this PR is to reduce the number of operations with storage.

Reply:
So, we do not need to do mathematical calculations, but you still need to do these calculations because there is no benefit from this? Given that these mathematical operations are strict with verification (checked)?

What exactly concerns this PR is that there is no need to update the cache. It's definitely a redundant operation.

Reply:
You just reduce 2 strict math operations, and 1 memory operation. Why it's not cheaper?

Reply:
> You just reduce 2 strict math operations, and 1 memory operation. Why it's not cheaper?

Because checking values is not free.

Reply:
1. Not free. but 4 math instructions are cheaper than 1 JUMP instruction and memory operation?!!
2. When using the Aurora+, zero gas operations are common. In this case, `update_cached_balance` is a guaranteed read of the Storage. And it is not needed in that particular case.